### DISCUSSION TITLE
Ensuring Reproducibility in Sequential Optuna Studies After Interruption

### DISCUSSION THREAD

**Discussion by csleo95** [optuna/optuna#5245] [Q&A] [answered]
Hello,

I'm using Optuna for hyperparameter optimization in a sequential, non-parallel execution environment. I've encountered an issue regarding the reproducibility of studies that are interrupted and then resumed.

When initializing the TPESampler with a fixed seed, I can achieve reproducible results across runs as long as the study is not interrupted. This is expected and works well. However, issues arise when an optimization process is interrupted (e.g., by a job scheduler) and subsequently resumed. Although I can use enqueue_trial to re-run the interrupted trial with the same parameters, ensuring that this particular trial is reproducible, the trials following the interrupted one diverge from their original parameter suggestions when the study was not interrupted, leading to a non-reproducible study from that point onwards.

### Steps to Reproduce

1. Start a study with TPESampler initialized with a specific seed.
2. Run the optimization process until a certain number of trials are completed.
3. Start another study with TPESampler initialized with the same seed.
4. Interrupt the optimization process before it reaches the number of trials of the first process.
5. Resume the study, using enqueue_trial for the interrupted trial to ensure its parameters match the pre-interruption state.
6. Continue the optimization process.

**Expected Result:** The study, when resumed, should produce the same parameter suggestions for all trials as it would have without interruption, ensuring complete reproducibility.

**Actual Result:** After resumption, parameter suggestions for trials following the interrupted one differ from an uninterrupted run, leading to reproducibility issues.

### Question
Is there a built-in mechanism or recommended practice in Optuna to handle such interruptions and ensure the reproducibility of the study?

Many thanks for your help, I really appreciate it!!

Best regards,
Leo

**Comment by nzw0301** [comment]
The issue comes from `study` not storing `sampler` instance(s). For reproducibility, we need to save and load sampler's too as explained in https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/001_rdb.html#resume-study.

**Reply by csleo95** [reply]
Thank you for sharing the documentation link; it's incredibly helpful! From what I understand, the process described works well when the optimization completes successfully, allowing the sampler to be saved and then reloaded for further optimization. However, what could be done if the study is interrupted before the sampler is saved? Is there a recommended approach for periodically saving the sampler's state during the optimization process to ensure reproducibility, even if an unexpected interruption occurs? Please, let me know if I didn't understand something correctly. Thanks for your help again!

**Reply by nzw0301** [reply, marked as answer]
You're right. I suppose one workaround is as follows:

1. Create a new study with the saved sampler that is saved before evaluating the interrupted trial.
2. Copy all trials except for the interrupted trial from an old study and to the new one using [`study.add_trials`](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html#optuna.study.Study.add_trials).
3. Resume the optimization.

**Reply by csleo95** [reply]
Thank you very much for the workaround! I'll try it!
